File: backend/src/main/java/com/aprendizaje/backend/service/TareaService.java
package com.aprendizaje.backend.service;

import com.aprendizaje.backend.model.Tarea;
import com.aprendizaje.backend.repository.TareaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Servicio que contiene la lógica de negocio para las Tareas
 * @Service indica que esta clase es un servicio de Spring
 */
@Service
public class TareaService {
    
    /**
     * Inyección de dependencias del repositorio
     * @Autowired indica que Spring debe inyectar automáticamente el repositorio
     */
    @Autowired
    private TareaRepository tareaRepository;
    
    /**
     * Obtiene todas las tareas de la base de datos
     * @return lista de todas las tareas
     */
    public List<Tarea> obtenerTodasLasTareas() {
        // Llama al método findAll() del repositorio
        return tareaRepository.findAll();
    }
    
    /**
     * Obtiene una tarea por su ID
     * @param id - identificador de la tarea
     * @return Optional con la tarea si existe, vacío si no existe
     */
    public Optional<Tarea> obtenerTareaPorId(Long id) {
        // Llama al método findById() del repositorio
        return tareaRepository.findById(id);
    }
    
    /**
     * Crea una nueva tarea en la base de datos
     * @param tarea - tarea a crear
     * @return la tarea creada con su ID generado
     */
    public Tarea crearTarea(Tarea tarea) {
        // Llama al método save() del repositorio
        return tareaRepository.save(tarea);
    }
    
    /**
     * Actualiza una tarea existente
     * @param id - identificador de la tarea a actualizar
     * @param tareaActualizada - datos actualizados de la tarea
     * @return la tarea actualizada, o null si no existe
     */
    public Tarea actualizarTarea(Long id, Tarea tareaActualizada) {
        // Verifica si la tarea existe
        Optional<Tarea> tareaExistente = tareaRepository.findById(id);
        
        if (tareaExistente.isPresent()) {
            // Si existe, actualiza los campos
            Tarea tarea = tareaExistente.get();
            tarea.setTitulo(tareaActualizada.getTitulo());
            tarea.setDescripcion(tareaActualizada.getDescripcion());
            tarea.setCompletada(tareaActualizada.getCompletada());
            
            // Guarda los cambios
            return tareaRepository.save(tarea);
        }
        
        // Si no existe, retorna null
        return null;
    }
    
    /**
     * Elimina una tarea por su ID
     * @param id - identificador de la tarea a eliminar
     * @return true si se eliminó, false si no existía
     */
    public boolean eliminarTarea(Long id) {
        // Verifica si existe
        if (tareaRepository.existsById(id)) {
            // Si existe, la elimina
            tareaRepository.deleteById(id);
            return true;
        }
        // Si no existe, retorna false
        return false;
    }
    
    /**
     * Marca una tarea como completada o no completada
     * @param id - identificador de la tarea
     * @param completada - true para marcar como completada, false para no completada
     * @return la tarea actualizada, o null si no existe
     */
    public Tarea cambiarEstadoTarea(Long id, Boolean completada) {
        // Busca la tarea
        Optional<Tarea> tareaExistente = tareaRepository.findById(id);
        
        if (tareaExistente.isPresent()) {
            // Si existe, cambia el estado
            Tarea tarea = tareaExistente.get();
            tarea.setCompletada(completada);
            
            // Guarda los cambios
            return tareaRepository.save(tarea);
        }
        
        // Si no existe, retorna null
        return null;
    }
}

File: frontend/src/services/tareaService.js
// URL base de la API
const API_URL = 'http://localhost:8080/api/tareas';

/**
 * Servicio para manejar todas las operaciones relacionadas con Tareas
 * Centraliza las llamadas HTTP al backend
 */
const tareaService = {
  
  /**
   * Obtiene todas las tareas del backend
   * @returns {Promise<Array>} Lista de tareas
   */
  obtenerTodas: async () => {
    try {
      const respuesta = await fetch(API_URL);
      if (!respuesta.ok) {
        throw new Error('Error al obtener las tareas');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en obtenerTodas:', error);
      throw error;
    }
  },

  /**
   * Crea una nueva tarea
   * @param {Object} tarea - Datos de la tarea a crear
   * @returns {Promise<Object>} La tarea creada
   */
  crear: async (tarea) => {
    try {
      const respuesta = await fetch(API_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(tarea)
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al crear la tarea');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en crear:', error);
      throw error;
    }
  },

  /**
   * Elimina una tarea
   * @param {number} id - ID de la tarea a eliminar
   * @returns {Promise<boolean>} true si se eliminó correctamente
   */
  eliminar: async (id) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}`, {
        method: 'DELETE'
      });
      
      if (!respuesta.ok && respuesta.status !== 204) {
        throw new Error('Error al eliminar la tarea');
      }
      
      return true;
    } catch (error) {
      console.error('Error en eliminar:', error);
      throw error;
    }
  },

  /**
   * Actualiza una tarea existente
   * @param {number} id - ID de la tarea a actualizar
   * @param {Object} tarea - Nuevos datos de la tarea
   * @returns {Promise<Object>} La tarea actualizada
   */
  actualizar: async (id, tarea) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(tarea)
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al actualizar la tarea');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en actualizar:', error);
      throw error;
    }
  },

  /**
   * Marca una tarea como completada
   * @param {number} id - ID de la tarea a completar
   * @returns {Promise<Object>} La tarea actualizada
   */
  completar: async (id) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}/completar`, {
        method: 'PUT'
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al completar la tarea');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en completar:', error);
      throw error;
    }
  },

  /**
   * Marca una tarea como no completada
   * @param {number} id - ID de la tarea a descompletar
   * @returns {Promise<Object>} La tarea actualizada
   */
  descompletar: async (id) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}/descompletar`, {
        method: 'PUT'
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al descompletar la tarea');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en descompletar:', error);
      throw error;
    }
  }
};

// Exportar el servicio para usarlo en otros archivos
export default tareaService;
